File: 0328-odd-even-linked-list/0328-odd-even-linked-list.kt
class Solution {
    fun oddEvenList(head: ListNode?): ListNode? {
        var oddHead = ListNode(-1)
        var evenHead = ListNode(-1)

        var i = head
        var odd = oddHead
        var even = evenHead
        var flag = true //true: odd, false: even

        while(i != null){
            if(flag) {
                odd.next = i
                odd = odd.next
            }
            else{
                even.next = i
                even = even.next
            } 
            i = i.next
            flag = !flag
        }
        odd.next = evenHead.next
        even.next = null
        return oddHead.next
    }
}

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function oddEvenList(head: ListNode | null): ListNode | null {
    const oddHead = new ListNode(0) //dummy head
    const evenHead = new ListNode(0) // dummy head

    let oddPrev = oddHead
    let evenPrev = evenHead
    let cur = head
    let isOdd = true
    while(cur){
        if(isOdd){ // add to odd list
            oddPrev.next = cur
            oddPrev = cur
        }
        else{ // add to even list
            evenPrev.next = cur
            evenPrev = cur
        }
        // move cur node
        isOdd = !isOdd
        cur = cur.next
    }
    oddPrev.next = evenHead.next // concat odd list & even list
    evenPrev.next = null // cut cycle
    return oddHead.next 
};